File: Project/Dropping.cpp
#include <iostream>
using namespace std;

int main() {
    int numTests;
    cout << "Enter the number of test cases: "<<endl;
    cin >> numTests;

    while (numTests--) {
        int deep, ball, ans = 1;
        cout << "Enter the depth and ball number: "<<endl;
        cin >> deep >> ball;

        for (int i = 1; i < deep; i++) {
            if (ball % 2) {
                ans = ans * 2;
                ball = (ball + 1) / 2;
            } else {
                ans = ans * 2 + 1;
                ball /= 2;
            }
        }
        
        cout << "Final position: " << ans << endl;
    }

    return 0;
}



// 5
// 4 2
// 3 4
// 10 1
// 2 2
// 8 128
// -1


// Enter the number of test cases: 5
// Enter the depth and ball number: 4 2
// Final position: 12
// Enter the depth and ball number: 3 4
// Final position: 7
// Enter the depth and ball number: 10 1
// Final position: 512
// Enter the depth and ball number: 2 2
// Final position: 3
// Enter the depth and ball number: 8 128
// Final position: 255

File: Project/Dropping.java
import java.util.Scanner;

public class Dropping {
    static int bound;

// Recursive function to calculate the final position of the ball
public static int getFinalPosition(int currentDepth, int currentBallNumber) {
    //System.out.println("Current depth: " + currentDepth + ", current ball number: " + currentBallNumber);
     //if the current depth is equal to or greater than the bound. If it is, it means the ball has reached the final depth, so the current depth is returned as the final position of the ball.
     if (currentDepth >= bound) {
        return currentDepth;
    }

    int leftChild = currentDepth * 2;
    int rightChild = leftChild + 1;
    System.out.println("Left child: " + leftChild + ", right child: " + rightChild);
    if (currentBallNumber % 2 == 0) {
        return getFinalPosition(rightChild, currentBallNumber / 2);
    } else {
        return getFinalPosition(leftChild, (currentBallNumber + 1) / 2);
    }
}

public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    System.out.print("Enter the number of test cases: ");
    int numTests = scanner.nextInt();
    System.out.print("Enter the depth and ball number: ");
    while (numTests-- > 0) {
        int depth = scanner.nextInt();
        int ballNumber = scanner.nextInt();
        bound = 1; // Initialize bound to 1

        for (int i = 0; i < depth - 1; i++) {
            bound *= 2; // Calculate the bound based on the given depth
        }

        int finalPosition = getFinalPosition(1, ballNumber); // Start from depth 1
        System.out.println("Final position: " + finalPosition);
    }
}

}
